### DISCUSSION TITLE
Hedera Liquidity Pool (LP) Service suggestion

### DISCUSSION THREAD

**Discussion by jamesgao319** [hiero-ledger/hiero-improvement-proposals#119] [Ideas]
We at Hashgraph Exchange (HEX) have been trying to implement an LP solution for the past three months. At this moment, Hedera Token Service (HTS) and Hedera Smart Contract Service (HSC) do not interact with each other, meaning we cannot handle transfer of HTS assets and Hbar within the HSC. Unless Hedera makes HTS and HSC interact with each other, it's difficult for DAPP developers to implement a liquidity pool on the hashgraph platform.

Here we suggest a simpler solution. Just like Hedera introduced HTS, we propose for Hedera to create a Hedera Liquidity Pool service. It is an essential step to implement decentralized finance (DeFi) on Hedera Hashgraph, which enables it to better compete with the Ethereum network.

**Comment by PimanWasHere** [comment]
Is the Liquidity Pool not a priority on the dev list?

**Comment by alzoo** [comment]
Totally agree, I've missed if this has been spoken about already in the town hall meetings, anyone got any info?

**Comment by anrid** [comment]
Wow, yeah I was wondering why there are no DEXs on Hedera yet. This needs to be addressed urgently for Hedera to have its DeFi revolution. Imho, we’re not going far without access to basic DeFi building blocks like DEXs, stable coins, borrowing/lending services, etc.

**Comment by kiraraneda** [comment]
This should be at top priority and would make an easy transition for many projects.

**Comment by djlongy** [comment]
Agreed, this would be a integral and necessary feature for DeFi solutions.

**Comment by parabhairava** [comment]
while focusing on the Enterprise sector & creating new use-cases for DLTs is great, the predominant use-case for DLTs today is DeFi. DeFi will soon be an indistinguishable part of what cryptocurrency means from first principles. with that said, the importance of such features cannot be understated,

**Comment by mattsmithies** [comment]
As someone that effectively created an early stage DEX and Pooling mechanisms on Hedera I would suggest that leaning into the strengths and speed of the network. I would personally adjust the perspective or belief that smart contracts are even required for us, as they have limited bandwidth within the network.

The abuse and financial risk of smart contracts are high [Vitalik even commented on this matter](https://cointelegraph.com/news/ethereum-must-innovate-beyond-just-dapps-for-defi-degens-vitalik-buterin). Hedera as a network provides us with the tools to create any solution, any appnet to drive these applications. 

There are parties, like Limechain and others building bridges, this is far more immediately important than DeFi or LP pools

Due to this realization (in February) this is the primary reason why more technical projects like **Unibar** were slowed down. 

_As a solution_ I would suggest exploring a new core Hedera service along the lines of a **Ledger** or **Exchange** service. 
Let's call it **HLS** Hedera Ledger Service, it could be in theory bundled into HTS.

The purpose of **HLS** would be to effectively extend accounts and HTS to provide state to issue token offers and execute atomic swaps, at the native layer. 

> A state store for the request and swap of HTS tokens.

The flow would be as follows:

1. A user wants to "offer" a token for sale.
2. They provide a **token_id** of their token and a criteria of expected outcomes, so I would like to sell (hedera) DOV for hedera bridged Bitcoin or Ethereum or Litecoin or anything. (This could solve the omni pool problem, provide Uniswap V3-like interfaces, and provide exponentially more opportunities for price routing)
3. A buyer could swap for a token based on the criteria, that has been offered on the native layer, price doesn't matter, only the quantity requested.

## A HLS client app

On the Application layer, it would be responsible for aggregating the liquidity options and matching up the different routing scenarios.

## Build the entire thing yourself

To be honest, any team could build this entire process out as a separate process as it isn't that difficult, all it would require is a validator network, think scheduled transactions for the upcoming auction demo.

All of these designs are always a little more complicated but my fear is that "duplicating" the experience of Ethereum on Hedera isn't a good idea, Hedera is a different beast entirely with a huge amount of potential that has to be unlocked. 

Personally, I'd love to see HLS become a thing as the network would have a decentralized mechanism for providing deep liquidity, imagine something like [woo network](https://woo.network/) at the native layer driven through developer friendly SDKs or APIs.

**Reply by kiraraneda** [reply]
Great write up and insight Matt.

**Reply by jamesgao319** [reply]
Great write up, Matt, thanks for sharing with us your insights. Totally agree with you on the limitations of the smart contract model. Are you currently working with Limechain on building bridges to Hedera?

**Reply by mattsmithies** [reply]
Hey @jamesgao319 message me your email on twitter/DM and I'll intro you to someone that will be able to help.

**Comment by gmaijoe** [comment]
I'm a bit shocked that there isn't more support for smart contracts on the Hedera platform.

If the hedera governing council will need to approve and implement internally every new feature, including something like the proposed Hedera Liquidity Pool service, that flies in the face of decentralization (which Hedera already has criticisms about) and speed of innovation.

Long story short, if developers cannot build meaningful DeFi apps easily on Hedera, these developers and talent will migrate to other high speed chains like Solana, Avalanche etc who do support.

On Solana, there's already tons of new permisionless tools like [Saber](https://saber.so/) (swaps), [Radium](https://raydium.io/swap/), [Sollet](https://sollet.io), [Serum](https://projectserum.com/#/) and countless more, all of these enabled by smart contracts.

Please note: i'm a long time supporter of Hedera Hashgraph and will continually support them as I'm a big fan of the consensus algorithm and team, but this is simply what I've noticed after spending time in the crypto space since 2014. I along with everyone else here wants to see Hedera succeed and this is simply my 2 cents.

**Comment by lucashenning** [comment]
As one of the [Hedera projects](https://hedera.com/users/infinite-by-suku), we support this proposal and we're interested in exploring a potential Hedera HLS solution for our [SUKU DeFi protocol](https://defi.suku.world/). 

Having a standardized way to manage liquidity on Hedera would be great! Ideally something like a "Uniswap-like protocol" that supports swaps and pooling on Hedera. A layer like this could also work as a price oracle for other services in the future.

As far as smart contracts are concerned, I agree with what has been said by @mattsmithies. Smart contracts may not be required for this. Implementing this as one of the core Hedera services would make this more robust, less risky, and easier to adopt as a new standard. 

Some of the potential requirements that come to mind:
- swap HTS/HBAR & HTS/HTS
- provide liquidity for HTS/HTS pair, earn interest for liquidity / fraction of tx fees based on pool share
- decentralized price oracle: calculate price based on pool balance

Our dev team will look into this and see if there's any way to support this from our end.

**Reply by jamesgao319** [reply]
Thank you for your support!

**Comment by BrianHBARbrain** [comment]
I agree on this proposal that smart contracts are not the direction needed.  Instead a core Hedera service would elevate the ecosystem.

Also creating one sided liquidity isn’t happening on other AMM is it ?
